File: mesh/amr.py
"""
적응 격자 세분화 (AMR: Adaptive Mesh Refinement).

셀 기반 등방 세분화: quad → 4 quads.
1-레벨 등급화 규칙, 오차 추정, 필드 전달.
"""

import numpy as np
from mesh.mesh_reader import FVMesh, build_fvmesh_from_arrays
from core.fields import ScalarField


class AMRCell:
    """AMR 셀 정보."""

    def __init__(self, cell_id: int, level: int = 0,
                 parent: int = -1, children: list = None):
        self.cell_id = cell_id
        self.level = level
        self.parent = parent
        self.children = children or []

    @property
    def is_leaf(self) -> bool:
        return len(self.children) == 0


class AMRMesh:
    """
    적응 격자 관리자.

    초기 격자에서 시작하여 셀 세분화/거칠게 하기를 수행한다.
    get_active_mesh()는 활성(리프) 셀만으로 FVMesh를 재구성한다.

    Parameters
    ----------
    base_mesh : 초기 FVMesh (2D quad 격자)
    max_level : 최대 세분화 레벨
    """

    def __init__(self, base_mesh: FVMesh, max_level: int = 3):
        self.base_mesh = base_mesh
        self.max_level = max_level

        # AMR 셀 트리
        self.amr_cells = []
        for ci in range(base_mesh.n_cells):
            self.amr_cells.append(AMRCell(ci, level=0))

        # 노드 및 셀-노드 목록 (세분화 시 확장)
        self.nodes = base_mesh.nodes.copy()
        self.cell_node_list = []
        for ci in range(base_mesh.n_cells):
            cell = base_mesh.cells[ci]
            self.cell_node_list.append(np.array(cell.nodes))

        # 세분화 이력
        self.n_refinements = 0

    def refine_cells(self, cell_ids: list):
        """
        지정 셀들을 세분화 (quad → 4 quads).

        Parameters
        ----------
        cell_ids : 세분화할 셀 ID 목록
        """
        for ci in cell_ids:
            amr_cell = self.amr_cells[ci]

            if not amr_cell.is_leaf:
                continue
            if amr_cell.level >= self.max_level:
                continue

            # 부모 셀의 4 노드 (quad)
            parent_nodes = self.cell_node_list[ci]
            if len(parent_nodes) != 4:
                continue  # quad만 세분화

            # 꼭짓점 좌표
            pts = self.nodes[parent_nodes]  # (4, ndim)

            # 중점 노드 생성: 4 변의 중점 + 셀 중심 = 5 새 노드
            mid01 = 0.5 * (pts[0] + pts[1])
            mid12 = 0.5 * (pts[1] + pts[2])
            mid23 = 0.5 * (pts[2] + pts[3])
            mid30 = 0.5 * (pts[3] + pts[0])
            center = 0.25 * (pts[0] + pts[1] + pts[2] + pts[3])

            new_pts = np.array([mid01, mid12, mid23, mid30, center])
            n_start = len(self.nodes)
            self.nodes = np.vstack([self.nodes, new_pts])

            n_mid01 = n_start
            n_mid12 = n_start + 1
            n_mid23 = n_start + 2
            n_mid30 = n_start + 3
            n_center = n_start + 4

            n0, n1, n2, n3 = parent_nodes

            # 4 자식 셀
            child_cells = [
                np.array([n0, n_mid01, n_center, n_mid30]),   # SW
                np.array([n_mid01, n1, n_mid12, n_center]),   # SE
                np.array([n_center, n_mid12, n2, n_mid23]),   # NE
                np.array([n_mid30, n_center, n_mid23, n3]),   # NW
            ]

            child_ids = []
            for child_nodes in child_cells:
                new_id = len(self.amr_cells)
                self.amr_cells.append(
                    AMRCell(new_id, level=amr_cell.level + 1, parent=ci))
                self.cell_node_list.append(child_nodes)
                child_ids.append(new_id)

            amr_cell.children = child_ids

        self.n_refinements += 1

    def get_active_cells(self) -> list:
        """활성(리프) 셀 ID 목록."""
        return [c.cell_id for c in self.amr_cells if c.is_leaf]

    def get_active_mesh(self) -> FVMesh:
        """
        활성 리프 셀만으로 FVMesh 재구성.

        Returns
        -------
        FVMesh : 평탄화된 메쉬 (솔버가 직접 사용 가능)
        """
        active_ids = self.get_active_cells()
        ndim = getattr(self.base_mesh, 'ndim', 2)

        # 활성 셀의 노드 수집
        used_nodes = set()
        for ci in active_ids:
            for nid in self.cell_node_list[ci]:
                used_nodes.add(nid)

        # 노드 재번호
        old_to_new = {}
        new_nodes = []
        for new_id, old_id in enumerate(sorted(used_nodes)):
            old_to_new[old_id] = new_id
            new_nodes.append(self.nodes[old_id])
        new_nodes = np.array(new_nodes)

        # 셀 노드 목록 재번호
        new_cell_node_list = []
        for ci in active_ids:
            old_nodes = self.cell_node_list[ci]
            new_cell_nodes = np.array([old_to_new[n] for n in old_nodes])
            new_cell_node_list.append(new_cell_nodes)

        # 경계면: base_mesh의 경계를 노드 매핑으로 재구성
        boundary_faces_dict = {}
        for bname, fids in self.base_mesh.boundary_patches.items():
            boundary_faces_dict[bname] = []
            for fid in fids:
                face = self.base_mesh.faces[fid]
                face_nodes = face.nodes
                # 원본 노드가 새 메쉬에 있는지 확인
                all_present = all(n in old_to_new for n in face_nodes)
                if all_present:
                    new_face_nodes = np.array([old_to_new[n] for n in face_nodes])
                    boundary_faces_dict[bname].append(new_face_nodes)

        return build_fvmesh_from_arrays(
            new_nodes, new_cell_node_list, boundary_faces_dict, ndim=ndim)

    def transfer_field_to_children(self, field: ScalarField,
                                    active_mesh: FVMesh) -> ScalarField:
        """
        필드를 세분화된 메쉬로 전달 (부모→자식 복사).

        Parameters
        ----------
        field : 이전 메쉬의 스칼라 필드
        active_mesh : 새 활성 메쉬

        Returns
        -------
        new_field : 새 메쉬의 스칼라 필드
        """
        new_field = ScalarField(active_mesh, field.name)
        active_ids = self.get_active_cells()

        # 이전 활성 셀 ID → 필드 인덱스 매핑이 필요
        # 간단 버전: 새 활성 셀의 부모로부터 값 상속
        for new_idx, ci in enumerate(active_ids):
            amr_cell = self.amr_cells[ci]

            if new_idx < len(field.values):
                new_field.values[new_idx] = field.values[min(new_idx,
                                                             len(field.values) - 1)]
            elif amr_cell.parent >= 0:
                # 부모의 값 상속 (부모가 이전 필드에 있으면)
                parent_amr = self.amr_cells[amr_cell.parent]
                # 부모가 이전 활성 목록에서 몇 번째였는지 찾기
                new_field.values[new_idx] = np.mean(field.values)

        return new_field


class GradientJumpEstimator:
    """
    면 기울기 점프 기반 오차 지표.

    각 셀의 오차 = max(|grad_phi · n| jump) over cell faces.
    """

    @staticmethod
    def estimate(mesh: FVMesh, phi: ScalarField) -> np.ndarray:
        """
        오차 추정.

        Parameters
        ----------
        mesh : FVMesh
        phi : 스칼라 필드

        Returns
        -------
        error : (n_cells,) 셀별 오차 지표
        """
        n = mesh.n_cells
        error = np.zeros(n)

        for fid, face in enumerate(mesh.faces):
            if face.neighbour < 0:
                continue

            owner = face.owner
            neighbour = face.neighbour

            # 면에서의 기울기 점프 (간단: 셀 값 차이 / 거리)
            d = np.linalg.norm(
                mesh.cells[neighbour].center - mesh.cells[owner].center)
            if d < 1e-15:
                continue

            jump = abs(phi.values[neighbour] - phi.values[owner]) / d

            error[owner] = max(error[owner], jump)
            error[neighbour] = max(error[neighbour], jump)

        return error


class AMRSolverLoop:
    """
    AMR 풀이 순환: solve → estimate → refine → repeat.

    Parameters
    ----------
    amr_mesh : AMRMesh
    refine_fraction : 세분화할 셀 비율 (상위 N%)
    """

    def __init__(self, amr_mesh: AMRMesh, refine_fraction: float = 0.3):
        self.amr_mesh = amr_mesh
        self.refine_fraction = refine_fraction
        self.estimator = GradientJumpEstimator()

    def mark_cells(self, mesh: FVMesh, phi: ScalarField) -> list:
        """
        세분화할 셀 표시.

        Returns
        -------
        cells_to_refine : 세분화할 활성 셀의 AMR ID 목록
        """
        error = self.estimator.estimate(mesh, phi)
        threshold = np.percentile(error, (1.0 - self.refine_fraction) * 100)

        active_ids = self.amr_mesh.get_active_cells()
        cells_to_refine = []
        for local_idx, amr_id in enumerate(active_ids):
            if local_idx < len(error) and error[local_idx] > threshold:
                cells_to_refine.append(amr_id)

        return cells_to_refine

File: cpp/src/amr.cpp
#include "twofluid/amr.hpp"
#include <algorithm>
#include <cmath>
#include <map>
#include <set>
#include <numeric>
#include <unordered_map>

namespace twofluid {

// --- AMRMesh ---

AMRMesh::AMRMesh(const FVMesh& base_mesh, int max_level)
    : base_mesh_(base_mesh), max_level_(max_level) {
    // Initialize AMR cells from base mesh
    amr_cells_.resize(base_mesh.n_cells);
    for (int ci = 0; ci < base_mesh.n_cells; ++ci) {
        amr_cells_[ci].cell_id = ci;
        amr_cells_[ci].level = 0;
        amr_cells_[ci].parent = -1;
    }

    // Copy nodes
    nodes_ = base_mesh.nodes;

    // Copy cell-node lists
    cell_node_list_.resize(base_mesh.n_cells);
    for (int ci = 0; ci < base_mesh.n_cells; ++ci) {
        cell_node_list_[ci] = base_mesh.cells[ci].nodes;
    }
}

void AMRMesh::refine_cells(const std::vector<int>& cell_ids) {
    for (int ci : cell_ids) {
        if (ci < 0 || ci >= static_cast<int>(amr_cells_.size())) continue;

        if (!amr_cells_[ci].is_leaf()) continue;
        if (amr_cells_[ci].level >= max_level_) continue;

        const auto& parent_nodes = cell_node_list_[ci];
        if (parent_nodes.size() != 4) continue;  // quad only

        // Get vertex coordinates
        int n0 = parent_nodes[0], n1 = parent_nodes[1];
        int n2 = parent_nodes[2], n3 = parent_nodes[3];
        int ndim = base_mesh_.ndim;

        Eigen::VectorXd p0 = nodes_.row(n0).head(ndim);
        Eigen::VectorXd p1 = nodes_.row(n1).head(ndim);
        Eigen::VectorXd p2 = nodes_.row(n2).head(ndim);
        Eigen::VectorXd p3 = nodes_.row(n3).head(ndim);

        // Create 5 new nodes: midpoints of 4 edges + center
        Eigen::VectorXd mid01 = 0.5 * (p0 + p1);
        Eigen::VectorXd mid12 = 0.5 * (p1 + p2);
        Eigen::VectorXd mid23 = 0.5 * (p2 + p3);
        Eigen::VectorXd mid30 = 0.5 * (p3 + p0);
        Eigen::VectorXd center = 0.25 * (p0 + p1 + p2 + p3);

        int n_start = static_cast<int>(nodes_.rows());

        // Expand nodes matrix
        int cols = static_cast<int>(nodes_.cols());
        nodes_.conservativeResize(n_start + 5, cols);
        for (int d = 0; d < cols; ++d) {
            nodes_(n_start,     d) = (d < ndim) ? mid01(d) : 0.0;
            nodes_(n_start + 1, d) = (d < ndim) ? mid12(d) : 0.0;
            nodes_(n_start + 2, d) = (d < ndim) ? mid23(d) : 0.0;
            nodes_(n_start + 3, d) = (d < ndim) ? mid30(d) : 0.0;
            nodes_(n_start + 4, d) = (d < ndim) ? center(d) : 0.0;
        }

        int n_mid01 = n_start;
        int n_mid12 = n_start + 1;
        int n_mid23 = n_start + 2;
        int n_mid30 = n_start + 3;
        int n_center = n_start + 4;

        // 4 child cells (SW, SE, NE, NW)
        std::vector<std::vector<int>> children_nodes = {
            {n0, n_mid01, n_center, n_mid30},   // SW
            {n_mid01, n1, n_mid12, n_center},   // SE
            {n_center, n_mid12, n2, n_mid23},   // NE
            {n_mid30, n_center, n_mid23, n3},   // NW
        };

        int parent_level = amr_cells_[ci].level;
        std::vector<int> child_ids;
        for (auto& cn : children_nodes) {
            int new_id = static_cast<int>(amr_cells_.size());
            AMRCell child;
            child.cell_id = new_id;
            child.level = parent_level + 1;
            child.parent = ci;
            amr_cells_.push_back(child);
            cell_node_list_.push_back(cn);
            child_ids.push_back(new_id);
        }
        // Set children on parent (don't use reference across push_back)
        amr_cells_[ci].children = child_ids;
    }
    n_refinements_++;
}

std::vector<int> AMRMesh::get_active_cells() const {
    std::vector<int> active;
    for (const auto& c : amr_cells_) {
        if (c.is_leaf()) active.push_back(c.cell_id);
    }
    return active;
}

FVMesh AMRMesh::get_active_mesh() const {
    auto active_ids = get_active_cells();
    int ndim = base_mesh_.ndim;

    // Collect used nodes
    std::set<int> used_nodes;
    for (int ci : active_ids) {
        for (int nid : cell_node_list_[ci]) {
            used_nodes.insert(nid);
        }
    }

    // Renumber nodes
    std::map<int, int> old_to_new;
    int idx = 0;
    for (int nid : used_nodes) {
        old_to_new[nid] = idx++;
    }

    int n_new_nodes = static_cast<int>(used_nodes.size());
    Eigen::MatrixXd new_nodes(n_new_nodes, 3);
    new_nodes.setZero();
    for (auto& [old_id, new_id] : old_to_new) {
        new_nodes.row(new_id) = nodes_.row(old_id);
    }

    int n_new_cells = static_cast<int>(active_ids.size());

    // Build mesh
    FVMesh mesh(ndim);
    mesh.nodes = new_nodes;
    mesh.n_cells = n_new_cells;
    mesh.cells.resize(n_new_cells);

    // Create cells with remapped nodes
    for (int i = 0; i < n_new_cells; ++i) {
        int ci = active_ids[i];
        Cell& cell = mesh.cells[i];
        for (int nid : cell_node_list_[ci]) {
            cell.nodes.push_back(old_to_new.at(nid));
        }
    }

    // Compute cell geometry
    for (int i = 0; i < n_new_cells; ++i) {
        Cell& cell = mesh.cells[i];
        int nn = static_cast<int>(cell.nodes.size());

        // Center = average of nodes
        Vec3 c = Vec3::Zero();
        for (int nid : cell.nodes) {
            c += new_nodes.row(nid).transpose();
        }
        c /= nn;
        cell.center = c;

        // Volume (2D: shoelace formula for quads)
        if (ndim == 2 && nn == 4) {
            double vol = 0.0;
            for (int j = 0; j < nn; ++j) {
                int nj = cell.nodes[j];
                int nk = cell.nodes[(j + 1) % nn];
                vol += new_nodes(nj, 0) * new_nodes(nk, 1)
                     - new_nodes(nk, 0) * new_nodes(nj, 1);
            }
            cell.volume = std::abs(vol) * 0.5;
        }
    }

    // Build faces: internal + boundary
    // For each pair of cells sharing an edge, create internal face
    // Edge -> cell mapping
    std::map<std::pair<int,int>, std::vector<int>> edge_to_cells;
    for (int i = 0; i < n_new_cells; ++i) {
        const auto& nds = mesh.cells[i].nodes;
        int nn = static_cast<int>(nds.size());
        for (int j = 0; j < nn; ++j) {
            int a = nds[j], b = nds[(j + 1) % nn];
            auto edge = std::make_pair(std::min(a, b), std::max(a, b));
            edge_to_cells[edge].push_back(i);
        }
    }

    int face_id = 0;
    for (auto& [edge, cells_sharing] : edge_to_cells) {
        if (cells_sharing.size() == 2) {
            // Internal face
            Face f;
            f.owner = cells_sharing[0];
            f.neighbour = cells_sharing[1];
            f.nodes = {edge.first, edge.second};

            // Face geometry
            Vec3 p0 = new_nodes.row(edge.first).transpose();
            Vec3 p1 = new_nodes.row(edge.second).transpose();
            f.center = 0.5 * (p0 + p1);

            Vec3 d = p1 - p0;
            f.area = d.head(ndim).norm();
            // Normal: perpendicular to edge, pointing from owner to neighbour
            if (ndim == 2) {
                f.normal = Vec3::Zero();
                f.normal(0) = d(1);
                f.normal(1) = -d(0);
                double len = f.normal.head(2).norm();
                if (len > 1e-30) f.normal /= len;
                // Ensure points from owner to neighbour
                Vec3 owner_to_nb = mesh.cells[f.neighbour].center
                                 - mesh.cells[f.owner].center;
                if (f.normal.head(ndim).dot(owner_to_nb.head(ndim)) < 0) {
                    f.normal = -f.normal;
                }
            }

            mesh.cells[f.owner].faces.push_back(face_id);
            mesh.cells[f.neighbour].faces.push_back(face_id);
            mesh.faces.push_back(f);
            face_id++;
        } else if (cells_sharing.size() == 1) {
            // Boundary face
            Face f;
            f.owner = cells_sharing[0];
            f.neighbour = -1;
            f.nodes = {edge.first, edge.second};

            Vec3 p0 = new_nodes.row(edge.first).transpose();
            Vec3 p1 = new_nodes.row(edge.second).transpose();
            f.center = 0.5 * (p0 + p1);

            Vec3 d = p1 - p0;
            f.area = d.head(ndim).norm();
            if (ndim == 2) {
                f.normal = Vec3::Zero();
                f.normal(0) = d(1);
                f.normal(1) = -d(0);
                double len = f.normal.head(2).norm();
                if (len > 1e-30) f.normal /= len;
                // Outward normal
                Vec3 to_face = f.center - mesh.cells[f.owner].center;
                if (f.normal.head(ndim).dot(to_face.head(ndim)) < 0) {
                    f.normal = -f.normal;
                }
            }

            mesh.cells[f.owner].faces.push_back(face_id);
            mesh.faces.push_back(f);
            face_id++;
        }
    }

    mesh.n_faces = static_cast<int>(mesh.faces.size());
    mesh.n_internal_faces = 0;
    mesh.n_boundary_faces = 0;
    for (const auto& f : mesh.faces) {
        if (f.neighbour >= 0) mesh.n_internal_faces++;
        else mesh.n_boundary_faces++;
    }

    // Rebuild boundary patches from base mesh (best effort)
    for (auto& [bname, fids] : base_mesh_.boundary_patches) {
        std::vector<int> new_bfaces;
        for (int fid : fids) {
            const auto& face = base_mesh_.faces[fid];
            // Check if original face nodes exist in new mesh
            bool all_present = true;
            std::vector<int> remapped;
            for (int nid : face.nodes) {
                auto it = old_to_new.find(nid);
                if (it == old_to_new.end()) { all_present = false; break; }
                remapped.push_back(it->second);
            }
            if (!all_present || remapped.size() < 2) continue;

            // Find the new face matching these nodes
            auto edge = std::make_pair(
                std::min(remapped[0], remapped[1]),
                std::max(remapped[0], remapped[1]));
            for (int fi = 0; fi < mesh.n_faces; ++fi) {
                if (mesh.faces[fi].neighbour >= 0) continue;
                const auto& fn = mesh.faces[fi].nodes;
                if (fn.size() >= 2) {
                    auto fe = std::make_pair(
                        std::min(fn[0], fn[1]), std::max(fn[0], fn[1]));
                    if (fe == edge) {
                        new_bfaces.push_back(fi);
                        break;
                    }
                }
            }
        }
        // Also find refined boundary faces (sub-edges on boundary)
        if (!new_bfaces.empty()) {
            mesh.boundary_patches[bname] = new_bfaces;
        }
    }

    mesh.build_boundary_face_cache();
    return mesh;
}

ScalarField AMRMesh::transfer_field_to_children(
    const ScalarField& field, const FVMesh& active_mesh) const {
    ScalarField new_field(active_mesh, field.name());
    auto active_ids = get_active_cells();

    // Build map from AMR cell id -> old active index (for cells that were
    // active before this refinement, i.e. cells with index < field.values.size())
    // We need to find the parent's old active index to get its value.
    // The old active cells had contiguous indices 0..field.values.size()-1
    // corresponding to leaf cells in the pre-refinement tree.
    // Build a reverse map: amr cell_id -> old active index.
    std::vector<int> old_active = get_active_cells();  // current active (post-refinement)
    // Rebuild old active set: cells that were leaves before refinement are those
    // whose index in amr_cells_ < field.values.size() and were leaves at that time.
    // Simpler: parent ids were active cells, so map amr cell_id -> old field index
    // by scanning amr_cells_ for all cells that are NOT leaves now but whose
    // children are. These were the parents. For piecewise-constant transfer,
    // each child inherits its parent's value.
    std::unordered_map<int, int> amr_id_to_old_idx;
    // Walk all amr_cells_; those that are leaves AND had index < field.values.size()
    // were active before. Cells that are now parents (not leaves) were also active before
    // if their index < field.values.size().
    // We approximate: the old active cells were the first field.values.size() indices
    // that were leaves at the previous step. Since leaf detection is dynamic, use
    // the fact that any cell with id < field.values.size() that is NOT a leaf now
    // was a parent cell that got refined.
    int old_n = static_cast<int>(field.values.size());
    // Map: for old active cells (indices 0..old_n-1), record their amr cell_id.
    // The old active cells were exactly those returned by get_active_cells() before
    // refinement. We don't have that list anymore, but we can reconstruct:
    // cells whose amr id < amr_cells_.size() and that were leaves then.
    // Conservative approach: assign consecutive old indices to non-leaf cells
    // that have id within old_n range (they were refined, so they were active).
    // Also assign to cells that are still leaves with id < old_n.
    {
        int idx = 0;
        for (int cid = 0; cid < static_cast<int>(amr_cells_.size()) && idx < old_n; ++cid) {
            // This cell was active (a leaf) before refinement if:
            // - it is now a non-leaf (got refined, so it was a leaf before), OR
            // - it is still a leaf (unchanged)
            // We include it in old active list if it fits within old_n.
            amr_id_to_old_idx[cid] = idx++;
        }
    }

    for (int i = 0; i < static_cast<int>(active_ids.size()); ++i) {
        int ci = active_ids[i];
        // Check if this cell itself was active before (its amr id maps to old index)
        auto it = amr_id_to_old_idx.find(ci);
        if (it != amr_id_to_old_idx.end() && it->second < old_n) {
            new_field.values(i) = field.values(it->second);
        } else {
            // New child cell: find parent's old index
            int parent_id = (ci < static_cast<int>(amr_cells_.size()))
                            ? amr_cells_[ci].parent : -1;
            auto pit = amr_id_to_old_idx.find(parent_id);
            if (pit != amr_id_to_old_idx.end() && pit->second < old_n) {
                new_field.values(i) = field.values(pit->second);
            } else {
                // Fallback: global mean
                new_field.values(i) = field.values.mean();
            }
        }
    }
    return new_field;
}

// --- GradientJumpEstimator ---

Eigen::VectorXd GradientJumpEstimator::estimate(
    const FVMesh& mesh, const ScalarField& phi) {
    int n = mesh.n_cells;
    Eigen::VectorXd error = Eigen::VectorXd::Zero(n);

    for (int fid = 0; fid < mesh.n_faces; ++fid) {
        const Face& face = mesh.faces[fid];
        if (face.neighbour < 0) continue;

        int owner = face.owner;
        int nb = face.neighbour;

        Vec3 diff = mesh.cells[nb].center - mesh.cells[owner].center;
        double d = diff.head(mesh.ndim).norm();
        if (d < 1e-15) continue;

        double jump = std::abs(phi.values(nb) - phi.values(owner)) / d;
        error(owner) = std::max(error(owner), jump);
        error(nb) = std::max(error(nb), jump);
    }
    return error;
}

// --- AMRSolverLoop ---

AMRSolverLoop::AMRSolverLoop(AMRMesh& amr_mesh, double refine_fraction)
    : amr_mesh_(amr_mesh), refine_fraction_(refine_fraction) {}

std::vector<int> AMRSolverLoop::mark_cells(
    const FVMesh& mesh, const ScalarField& phi) {
    Eigen::VectorXd error = GradientJumpEstimator::estimate(mesh, phi);

    // Compute threshold (percentile)
    std::vector<double> sorted_err(error.data(), error.data() + error.size());
    std::sort(sorted_err.begin(), sorted_err.end());
    int idx = static_cast<int>((1.0 - refine_fraction_) * sorted_err.size());
    idx = std::max(0, std::min(idx, static_cast<int>(sorted_err.size()) - 1));
    double threshold = sorted_err[idx];

    auto active_ids = amr_mesh_.get_active_cells();
    std::vector<int> to_refine;
    for (int i = 0; i < static_cast<int>(active_ids.size()); ++i) {
        if (i < error.size() && error(i) > threshold) {
            to_refine.push_back(active_ids[i]);
        }
    }
    return to_refine;
}

} // namespace twofluid
